From 926eb43fa06b3b2e4605eb55e4a113aa997abf93 Mon Sep 17 00:00:00 2001
From: Hanna Hawa <hannah@marvell.com>
Date: Tue, 12 Jul 2016 12:55:55 +0300
Subject: [PATCH 0615/2241] fix: mmc: enable slow mode for SDHCI-XENON driver

This mode is needed when working with low frequencies (<50MHz)

Change-Id: Ia34d849735d92cbb8762d51e29563be630aa56fc
Signed-off-by: Hanna Hawa <hannah@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/31065
Reviewed-by: Victor Gu <xigu@marvell.com>
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Neta Zur Hershkovits <neta@marvell.com>
---
 drivers/mmc/host/sdhci-xenon.c | 31 +++++++++++++------------------
 1 file changed, 13 insertions(+), 18 deletions(-)

diff --git a/drivers/mmc/host/sdhci-xenon.c b/drivers/mmc/host/sdhci-xenon.c
index 226c892..8048566 100644
--- a/drivers/mmc/host/sdhci-xenon.c
+++ b/drivers/mmc/host/sdhci-xenon.c
@@ -309,25 +309,20 @@ static void sdhci_xenon_phy_reset(struct sdhci_host *host, unsigned int timing)
 		sdhci_writel(host, reg, EMMC_PHY_TIMING_ADJUST);
 	}
 
-	/* If SDIO card, set SDIO Mode
-	 * Otherwise, clear SDIO Mode and Slow Mode
-	 */
-	if (card && (card->type & MMC_TYPE_SDIO)) {
-		reg = sdhci_readl(host, EMMC_PHY_TIMING_ADJUST);
+	reg = sdhci_readl(host, EMMC_PHY_TIMING_ADJUST);
+	/* Clear SDIO and slow modes */
+	reg &= ~(TIMING_ADJUST_SDIO_MODE | TIMING_ADJUST_SLOW_MODE);
+	/* If SDIO card, set SDIO Mode */
+	if (card && (card->type & MMC_TYPE_SDIO))
 		reg |= TIMING_ADJUST_SDIO_MODE;
-
-		if ((timing == MMC_TIMING_UHS_SDR25) ||
-		    (timing == MMC_TIMING_UHS_SDR12) ||
-		    (timing == MMC_TIMING_SD_HS) ||
-		    (timing == MMC_TIMING_LEGACY))
-			reg |= TIMING_ADJUST_SLOW_MODE;
-
-		sdhci_writel(host, reg, EMMC_PHY_TIMING_ADJUST);
-	} else {
-		reg = sdhci_readl(host, EMMC_PHY_TIMING_ADJUST);
-		reg &= ~(TIMING_ADJUST_SDIO_MODE | TIMING_ADJUST_SLOW_MODE);
-		sdhci_writel(host, reg, EMMC_PHY_TIMING_ADJUST);
-	}
+	/* Enable slow mode in SDR mode below 50M */
+	if ((timing == MMC_TIMING_UHS_SDR50) ||
+	    (timing == MMC_TIMING_UHS_SDR25) ||
+	    (timing == MMC_TIMING_UHS_SDR12) ||
+	    (timing == MMC_TIMING_SD_HS) ||
+	    (timing == MMC_TIMING_LEGACY))
+		reg |= TIMING_ADJUST_SLOW_MODE;
+	sdhci_writel(host, reg, EMMC_PHY_TIMING_ADJUST);
 
 	/* Set preferred ZNR and ZPR value
 	 * The ZNR and ZPR value vary between different boards.
-- 
2.7.4

